includes/enrolment/class-sensei-enrolment-provider-journal-store.php

```diff
@@ -0,0 +1,283 @@
+		$journal_store = self::get( $user_id, $course_id );
+
+		$has_changed = false;
+
+		foreach ( $provider_results as $provider_id => $is_enrolled ) {
+			if ( ! isset( $journal_store->providers_journal[ $provider_id ] ) ) {
+				$journal_store->providers_journal[ $provider_id ] = Sensei_Enrolment_Provider_Journal::create();
+			}
+
+			$has_changed = $journal_store->providers_journal[ $provider_id ]->update_enrolment_status( $is_enrolled ) || $has_changed;
+		}
+
+		$current_snapshot = self::get_enrolment_snanpshot( $user_id, $course_id );
+
+		$removed_providers = array_diff( array_keys( $current_snapshot ), array_keys( $provider_results ) );
+		foreach ( $removed_providers as $removed_provider ) {
+			$has_changed = $journal_store->providers_journal[ $removed_provider ]->delete_enrolment_status() || $has_changed;
+		}
+
+		$journal_store->has_changed = $has_changed;
+	}
+
+	/**
+	 * Get a snapshot of the providers' enrolment status for a user and course at a specific timestamp. If no timestamp
+	 * is provided the current snapshot is returned.
+	 *
+	 * @param int $user_id     The user to return the snapshot for.
+	 * @param int $course_id   The course to return the snapshot for.
+	 * @param int $timestamp   The timestamp of the snapshot in microseconds. If omitted the current snapshot is returned.
+	 *
+	 * @return array
+	 */
+	public static function get_enrolment_snanpshot( $user_id, $course_id, $timestamp = null ) {
+		$timestamp     = null === $timestamp ? microtime( true ) : $timestamp;
+		$journal_store = self::get( $user_id, $course_id );
+
+		$snapshot = [];
+		foreach ( $journal_store->providers_journal as $provider_id => $journal ) {
+			$status = $journal->get_status_at( $timestamp );
+
+			if ( null !== $status['enrolment_status'] ) {
+				$snapshot[ $provider_id ] = $status['enrolment_status'];
+			}
+		}
+
+		return $snapshot;
+	}
+
+	/**
+	 * Get the enrolment status history of a provider for a specified user and course.
+	 *
+	 * @param int $user_id     The user id.
+	 * @param int $course_id   The course id.
+	 * @param int $provider_id The provider id.
+	 *
+	 * @return array The history of the provider. Each element of the array has the format:
+	 *               [ 'timestamp' => Timestamp of the status change, 'enrolment_status' => true|false|null ]
+	 */
+	public static function get_provider_history( $user_id, $course_id, $provider_id ) {
+		$journal_store = self::get( $user_id, $course_id );
+
+		return isset( $journal_store->providers_journal[ $provider_id ] ) ? $journal_store->providers_journal[ $provider_id ]->get_history() : [];
+	}
+
+	/**
+	 * Add a log message to a provider.
+	 *
+	 * @param int    $user_id     The user id.
+	 * @param int    $course_id   The course id.
+	 * @param int    $provider_id The provider id.
+	 * @param string $message     The message to be added.
+	 */
+	public static function add_provider_log_message( $user_id, $course_id, $provider_id, $message ) {
```

Comment:
I was wondering if we could pass the provider object instead of `$provider_id` in these log related methods. If we do, would it make more sense to make it the first argument? Since these are interfaced, it might allow us to be flexible in how/when we store these in the future.

Reply:
Fixed! a9021c69ea1e36e922d87fdd85042b20a67b8e2c